lib-alert-tree: Convert children annotation from coma separated promql to JSON Path.

JBWatenbergScality opened the pull request:
**Component**:

lib-alert-tree

**Context**: 

In order to ease parsing and alert tree building by consumers we decided to switch
from a non standard coma separated set of promql queries defining the child alerts
to a well defined standard JSON Path.

JBWatenbergScality committed:
lib-alert-tree: Convert children annotation from coma separated promql to JSON Path.

In order to ease parsing and alert tree building by consumers we decided to switch
from a non standard coma separated set of promql queries defining the child alerts
to a well defined standard JSON Path.

gdemonet commented on tools/lib-alert-tree/lib_alert_tree/models.py:
```diff
@@ -156,8 +161,9 @@ def build_tree(self, parent=None, tree=None):
 
     @property
     def alert_rule(self):
+        json_path_filters = " || ".join(map(lambda child: child.json_path, self.children))
```
This depends on the selected `Relationship`, so we should implement the "JSONPath query" in there.

gdemonet commented on tools/lib-alert-tree/lib_alert_tree/prometheus.py:
```diff
@@ -86,6 +98,12 @@ def child_id(self):
 
+    @property
+    def child_json_path(self):
+        """A set of JSON Path filters representing this alert, for use in annotations."""
```
Technically not a set, rather an expression (since they are `&&`-ed together).

Also, "use in annotations" is not really describing the intent. Maybe reword to:

```suggestion
        """A JSONPath filter expression for selecting this alert as a child.
        
        This expression will be combined into a full JSONPath query for retrieving
        all children of a derived alert, exposed in an annotation for consumption
        by clients (such as UIs).
        """
```

gdemonet replied:
(granted, it's almost the same docstring as just above for `child_id`)

gdemonet commented on tools/lib-alert-tree/lib_alert_tree/models.py:
```diff
@@ -156,8 +161,9 @@ def build_tree(self, parent=None, tree=None):
+        json_path_filters = " || ".join(map(lambda child: child.json_path, self.children))
         self.params.setdefault("annotations", {}).update(
-            {"children": ", ".join(map(str, self.children))}
+            {"children": f"$[?({json_path_filters})]" }
```
Wondering if we couldn't "simply" add a key for this, since the other one provides a more user-readable syntax :thinking:

gdemonet commented on tools/lib-alert-tree/lib_alert_tree/models.py:
```diff
@@ -116,6 +121,15 @@ def build_query(self, children, group_by=None):
+        elif self == self.ANY:
+            join_by = " || "
+        json_path_filters = join_by.join(map(lambda child: child.json_path, children))
+        return f"$[?({json_path_filters})]"
```
Now that's implement here, I realize we're missing the "group by" logic that we have in the `Relationship.build_query` method.

The idea behind this logic is the following:

- the child query is used to filter the domain of alerts
- the `group_by` argument then decomposes the filtered domain into units which will become alerts

In practice:

- say you have a `Parent` alert with a single child `Child{kind='example'}`, and its `group_by` constructor argument set to `["name"]`
- assume the `Child` alert has two labels, `kind` and `name`
- when building the subtree for a `Parent` instance, you'll want to also filter the `Child` alerts on `name`

This means we should propagate the `group_by` attribute to `prometheus.AlertRule` objects, so we can render it in another annotation and as part of the JSONPath query.

JBWatenbergScality replied:
done in d28eb95

gdemonet commented on tools/lib-alert-tree/lib_alert_tree/prometheus.py:
```diff
@@ -86,6 +98,16 @@ def child_id(self):
+
+        This expression will be combined into a full JSONPath query for retrieving
+        all children of a derived alert, exposed in an annotation for consumption
+        by clients (such as UIs)."""
```
[PEP 257 - Docstring Conventions](https://www.python.org/dev/peps/pep-0257/) recommends the following style:

```suggestion
        by clients (such as UIs).
        """
```

gdemonet commented on tools/lib-alert-tree/tests/test_models.py:
```diff
@@ -101,12 +101,22 @@ def test_all():
+        assert (
+            models.Relationship.ALL.build_json_path([models.ExistingAlert("test")])
+            == "$[?((@.labels.alertname === 'test'))]"
+        )
```
Could be interesting to refactor these tests with `pytest.mark.parametrize`, but it's not urgent. Will open a debt ticket.

JBWatenbergScality committed:
test: Parametrized Relationship tests

JBWatenbergScality committed:
alert-tree: Add missing sum in prom rules expression so that a single parent alert is raised when several children are

bert-e commented:
# Hello jbwatenbergscality,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @chengyanjin

* @nicolast

* @jbwatenbergscality

* @teddyandrieux

* @lucieleonard





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

JBWatenbergScality committed:
lib-alert-tree: Add group by label filters to childrenJsonPath annotation

Some Parent alerts may be deduplicated for by components instances (e.g a VolumeAtRisk alert is raised for each pvc). To build a correct tree representation we then need to filter the child alerts matching those additional labels.
When defining an alert through a Prometheus rule, we can use a templating language in annotations to retrieve labels values of the expression that raised the alert.
In this commit we are using this templating language to make a reference to build a filter that only matches responsible children for a logical one.

JBWatenbergScality commented:
Regarding parametrised tests they weren't collected by pytest despite the `@staticmethod` annotation so I had to drop the class and I renamed the 2 tests to make it explicit that they are testing the Relationship behavior.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @chengyanjin

* @nicolast

* @jbwatenbergscality

* @teddyandrieux

* @lucieleonard

gdemonet commented on tools/lib-alert-tree/tests/test_models.py:
```diff
@@ -91,52 +91,89 @@ def test_existing_alert_severity_check():
+            "sum by (instance) (ALERTS{alertname='test1', alertstate='firing'} and "
+            "ALERTS{alertname='test2', alertstate='firing'}) >= 1",
+            "$[?(((@.labels.alertname === 'test1') && "
+            "(@.labels.alertname === 'test2')) && "
```
Actually, not sure how this filter expression can work, since no alert can honor `(@.labels.alertname === 'test1') && (@.labels.alertname === 'test2')`

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

* @chengyanjin

* @nicolast

* @jbwatenbergscality

* @teddyandrieux

* @lucieleonard





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

gdemonet commented on tools/lib-alert-tree/lib_alert_tree/models.py:
```diff
@@ -112,10 +117,26 @@ def build_query(self, children, group_by=None):
+    @staticmethod
+    def build_json_path(children, group_by=None):
+        """Build a JSON Path to retrieve alert children."""
+        json_path_filters = " || ".join(map(lambda child: child.json_path, children))
```
`map(lambda x: ..., iterable)` is not idiomatic Python, we usually prefer to use generator expressions:

```suggestion
        json_path_filters = " || ".join(child.json_path for child in children)
```

gdemonet commented on tools/lib-alert-tree/lib_alert_tree/models.py:
```diff
@@ -112,10 +117,26 @@ def build_query(self, children, group_by=None):
+                    lambda key: f"@.labels.{key} === '{{{{ $labels.{key} }}}}'",
+                    group_by,
+                )
+            )
```
Same here:

```suggestion
            group_filters = " && ".join(
                f"@.labels.{key} === '{{{{ $labels.{key} }}}}'"
                for key in group_by
            )
```

gdemonet commented on tools/lib-alert-tree/lib_alert_tree/models.py:
```diff
@@ -157,7 +178,12 @@ def build_tree(self, parent=None, tree=None):
         self.params.setdefault("annotations", {}).update(
-            {"children": ", ".join(map(str, self.children))}
+            {
+                "children": ", ".join(map(str, self.children)),
```
Note: here, we use map, because we only apply a pre-existing function. It could feel more verbose written as `str(child) for child in self.children`, so I personally like to use `map(str, self.children)` instead. Matter of preference, I'd say.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @chengyanjin

* @nicolast

* @lucieleonard

* @teddyandrieux

* @jbwatenbergscality

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @chengyanjin

* @nicolast

* @lucieleonard

* @teddyandrieux

* @jbwatenbergscality

* @alexandre-allard-scality

* @thomasdanan

* @gdemonet

JBWatenbergScality commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.11`



The following branches will **NOT** be impacted:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.11`



The following branches have **NOT** changed:

* `development/2.0`
* `development/2.1`
* `development/2.10`
* `development/2.2`
* `development/2.3`
* `development/2.4`
* `development/2.5`
* `development/2.6`
* `development/2.7`
* `development/2.8`
* `development/2.9`



Please check the status of the associated issue None.

Goodbye jbwatenbergscality.
